crates/cli/src/suite.rs

```diff
@@ -0,0 +1,139 @@
+//! Group benchmarks into suites.
+//!
+//! [Suite]s are files that contain a newline-separated list of benchmark files
+//! to run; see [Suite::parse] for more details on the syntax of these files. To
+//! distinguish between a [Suite] file and a regular benchmark file, we use
+//! [BenchmarkOrSuite].
+
+use anyhow::Result;
+use std::{
+    ffi::OsStr,
+    fs, io,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+
+/// Decide between a suite of benchmarks or an individual benchmark file.
+#[derive(Debug)]
+pub enum BenchmarkOrSuite {
+    Suite(Suite),
+    Benchmark(PathBuf),
+}
+
+impl BenchmarkOrSuite {
+    /// List all of paths to the benchmarks to run.
+    pub fn paths(&self) -> Vec<String> {
+        match self {
+            BenchmarkOrSuite::Suite(suite) => suite
+                .benchmarks
+                .iter()
+                .map(|p| p.display().to_string())
+                .collect(),
+            BenchmarkOrSuite::Benchmark(path) => vec![path.display().to_string()],
+        }
+    }
+}
+
+/// It is helpful to reference the original path.
+impl AsRef<OsStr> for BenchmarkOrSuite {
+    fn as_ref(&self) -> &OsStr {
+        match self {
+            BenchmarkOrSuite::Suite(suite) => suite.path.as_os_str(),
+            BenchmarkOrSuite::Benchmark(path) => path.as_os_str(),
+        }
+    }
+}
+
+/// Parse a [BenchmarkOrSuite] from a string path; files ending in `.suite` are
+/// assumed to be suite files.
+impl FromStr for BenchmarkOrSuite {
+    type Err = anyhow::Error;
+    fn from_str(path: &str) -> Result<Self, Self::Err> {
+        Ok(if Suite::has_extension(path) {
+            Self::Suite(Suite::parse(path)?)
+        } else {
+            Self::Benchmark(PathBuf::from(path))
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct Suite {
+    path: PathBuf,
+    benchmarks: Vec<PathBuf>,
+}
+
+impl Suite {
+    /// Parse the contents of a suite file:
+    /// - empty lines are ignored
+    /// - `#`-prefixed lines are ignored
+    /// - relative paths are resolved using the parent directory of the
+    ///   `suite_path`.
+    pub fn parse<P: AsRef<Path>>(suite_path: P) -> Result<Self> {
+        Self::parse_contents(suite_path.as_ref(), &fs::read(suite_path.as_ref())?)
+    }
+
+    /// Utility function for easier testing.
+    fn parse_contents<P: AsRef<Path>>(suite_path: P, file_contents: &[u8]) -> Result<Self> {
+        use io::BufRead;
+        let suite_path = suite_path.as_ref().to_path_buf();
+        let parent_dir = suite_path
+            .parent()
+            .expect("the suite path must have a parent directory");
```

Comment:
I think this should return an `Err` rather than panic.

Reply:
If a suite is a file, it should be contained in a directory, right? I thought this error case was too uncommon to force the caller to figure out what to do with it?

Reply:
Ah okay I misunderstood.

Maybe clarify in the message something like

> suites are always files and files always have parents

and also `debug_assert!(suite_path.is_file())` or something since this function expects that and that invariant is maintained by callers rather than this function itself.

Reply:
(Had to `debug_assert!` in `parse` instead of `parse_contents` to avoid creating files during testing... seems like an OK trade-off).